### PR TITLE
Add SnakeCasePropertyNamesContractResolver.

### PR DESCRIPTION
**Pull request by StephenCleary**

### CODE REVIEW

File: tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+using NUnit.Framework;
+using Newtonsoft.Json;
+
+namespace Faithlife.Json.Tests
+{
+	[TestFixture]
+	public class SnakeCasePropertyNamesContractResolverTests
+	{
+		[Test]
+		public void UsingStringComparisons()
```
Comment (ejball) [inline tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs:12]: What does this test name mean? Maybe `DeserializeModel`?

File: tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		{
+			var result = JsonConvert.DeserializeObject<Model>(
```
Comment (ejball) [inline tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs:14]: Does the contract resolver affect serialization as well? Should we test it?

File: src/Faithlife.Json/ContractResolvers/SnakeCasePropertyNamesContractResolver.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+	public sealed class SnakeCasePropertyNamesContractResolver : DefaultContractResolver
+	{
+		protected override string ResolvePropertyName(string propertyName)
+		{
+			return s_snakeCaseRegex.Replace(propertyName,
+				match => (match.Groups[1].Success ? match.Groups[1] + "_" : "") + match.Groups[2].ToString().ToLowerInvariant());
+		}
+
+		private static readonly Regex s_snakeCaseRegex = new Regex(@"([a-z])?([A-Z]+)");
```
Comment (ejball) [inline src/Faithlife.Json/ContractResolvers/SnakeCasePropertyNamesContractResolver.cs:17]: Should we handle numbers?

FWIW, here's how Facility does [snake case](https://github.com/FacilityApi/Facility/blob/5d5073737726323b18dc517feccd2a65b9aea40c/src/Facility.Definition/CodeGen/CodeGenUtility.cs#L53-L64) ([tests](https://github.com/FacilityApi/Facility/blob/63c8a822a9d8458fffe12d3b222461631a45fd2b/tests/Facility.Definition.UnitTests/CodeGen/CodeGenUtilityTests.cs#L47-L56)). It doesn't merge acronyms into the following word.

File: tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+				@"{
+					""regular_case"": 7,
+					""with_acronym_at_end"": 11,
+					""withacronym_at_beginning"": 13,
```
Comment (ejball) [inline tests/Faithlife.Json.Tests/SnakeCasePropertyNamesContractResolverTests.cs:18]: Are we sure about this? I get that we want `OAuth` to be `oauth`, but does this work (or not work) for other examples?

**Comment by bgrainger** [comment]
How is this different from the built-in [`SnakeCaseNamingStrategy`](https://www.newtonsoft.com/json/help/html/NamingStrategySnakeCase.htm)?

Assuming it is different, and needs to exist for a particular reason, please add XML documentation to the class explaining its rationale.

**Comment by StephenCleary** [comment]
@bgrainger I totally missed that this is built-in now! When I searched (for a snake-case *contract resolver*), I only found an old issue where James said he wasn't going to include it in JSON.NET. Guess that changed!

**Pull request closed without merging**